Reduce crafting proficiency penalties based on partial proficiency level

eltank opened the pull request:
#### Summary
Balance "Reduce crafting proficiency penalties based on partial proficiency level"


#### Purpose of change

As promised in #48650 here is a PR that implements one of @I-am-Erk's ideas from #44680:

> Once we have the ability to reduce proficiency penalties by a factor, the same infrastructure could and should be used to make proficiency penalties gradually fall off as you get close to the necessary xp to have the proficiency.


#### Describe the solution

Add `get_proficiency_level(proficiency_id)` to the Character class. It returns a number between 0 and 1 based on the progress made towards acquiring that proficiency.
Modify `recipe::proficiency_time_maluses()` and `recipe::proficiency_failure_maluses()` to reduce the malus for a given proficiency based on the value from `get_proficiency_level()`. Time malus mitigation and failure malus mitigation are computed using separate formulas. The failure penalty is not completely eliminated until the proficiency is mastered.

![cos](https://user-images.githubusercontent.com/8000047/121106799-96ef2980-c7bb-11eb-87fc-9a5f568eae5d.png)

![cube](https://user-images.githubusercontent.com/8000047/120943357-342a5f00-c6e3-11eb-97a1-1094f3b1e4c6.png)

#### Describe alternatives you've considered

Leaving things alone until the formula gets reworked post-0.F. But I think the current behavior of binary proficiency (either you know it or you don't) is unrealistic and should be fixed ~~before 0.F~~ [well, looks like that's not going to happen]

#### Testing

Added a unit test asserting that time taken to craft a recipe with partial proficiency is between the time taken with full proficiency and the time taken with no proficiency.
Playtesting: in a fresh debug world, boost skills to 10, learn all recipes, debug life support and spawn lots of planks and long strings. Then start crafting survival bows one at a time. ~~Each one takes less time than the one before.~~ At first there is barely any change, after about 20% proficiency you start seeing very small reductions in the time and failure modifiers, then after 50% the reductions become more significant. It takes over a day of crafting 25 survivor bows to achieve proficiency in Bowyery.

#### Additional context

It's possible that this may unbalance the crafting malus adjustments made in #49046, since that relied on playtesting data obtained in the absence of this change. However, given the recent player backlash about the impact of missing proficiencies on crafting (#48981) I don't think anyone is going to complain about crafting becoming a bit easier.

eltank commented:
@I-am-Erk I have two questions for you:
- Should the crafting malus mitigation based on partial proficiency use a linear formula (x% proficiency results in malus being reduced to x% of its max value) or would you rather use something else (e.g. polynomial, exponential, quadratic, sigmoid)?
- If a helper NPC is more proficient than you, should their proficiency level be used in calculations instead of yours? If they're fully proficient, this is already the case (your missing proficiency is fully mitigated).

I-am-Erk commented:
> * Should the crafting malus mitigation based on partial proficiency use a linear formula (x% proficiency results in malus being reduced to x% of its max value) or would you rather use something else (e.g. polynomial, exponential, quadratic, sigmoid)?

I think it should begin to kick in at 50% and probably be exponential. There should also be a boon when you hit 100%, 99% should still be a good bit less. So, something like:

`mitigation = ( prof_percent - 0.5, min 0 )^2`

> * If a helper NPC is more proficient than you, should their proficiency level be used in calculations instead of yours? If they're fully proficient, this is already the case (your missing proficiency is fully mitigated).

I don't see why not. We can basically assume that they're doing it and showing you how if you're nonproficient, or giving you prompts and helping prepare things for you if you're partially proficient.

eltank commented:
Here is what your proposed formula looks like:
![square](https://user-images.githubusercontent.com/8000047/120915667-ecabc080-c659-11eb-800c-bfa1ad407774.png)

Do you really want the mitigation to cap off at 25% when the player's proficiency is close to 100%? That seems a bit low.

You mentioned exponential, but the example you gave is polynomial. Here's an alternative formula that is exponential:
![exp100](https://user-images.githubusercontent.com/8000047/120915722-43b19580-c65a-11eb-9382-cc4a32e579c4.png)
This one is calibrated to give maximum 50% mitigation in the limit as proficiency level approaches 100%, but that can be scaled to any value by changing the denominator. Also the curve can be made steeper by increasing the base.

Alternatively, here's another polynomial curve with a higher exponent
![quad](https://user-images.githubusercontent.com/8000047/120916068-32698880-c65c-11eb-9397-a48f7e7f24be.png)
This one can be made steeper by increasing the exponent.

Due to the difficulty of setting player proficiency levels to specific values in unit tests, I'd prefer to use a continuous formula that gives tiny gains even at low proficiency values (like the last 2 examples).

Since this is tagged with "0.F Feature Freeze", does it mean you're against including this in 0.F?

I-am-Erk commented:
Either of your other formulae there work fine yep. I have no strong preference, mitigation up to 50% and a smoother transition with an inflection point somewhere around halfway to proficiency would balance out nicely.

I put the feature freeze label on this because it's a fairly significant balance shift that should wait until 0.F rolls out, it could potentially introduce new behaviours and bugs that need fixing and it's too late for that for stable. I believe there are ways to pre-merge it to 0.G when it's ready.

eltank commented:
Ok, I will need to rebase this onto 0.F-dev, but in order to do that I need to wait until my earlier PR #48650 gets merged into that branch. I don't have any visibility into the process but I think it will take a few days. I'm going to hold off on making any more code changes until then because rebasing a PR gets harder as you accumulate more commits / diff-LOC.

In the mean time, I have a few more thoughts on the theory behind this change that I'd like to discuss.
I believe there are 2 kinds of factors that affect proficiency maluses:
A) Routine memorization (a.k.a. "muscle memory"). This is acquired gradually via repetition, with no discontinuities.
B) Acquisition of information (e.g. "don't do X", "always to Y before Z", "use this much force when doing T"). These are things that you either know or don't, so the accumulation of all necessary knowledge in order to become proficient will have discontinuities (jumps).
I think the time malus is primarily influenced by A, while the failure malus is primarily influenced by B. Therefore I think the formula for mitigating the time malus should not be the same as the formula for mitigating the failure malus.

Consider the following thought experiment. You're already proficient in knitting but want to get faster at it (speed knitting). You spend 19 hours, 59 minutes and 59 seconds knitting scarfs. Do you think that spending one more second knitting is suddenly going to double your speed? I think not. The more likely scenario is that you've already gotten as fast as you're going to get and after 20 hours you can't improve your speed any more. 
For this reason I believe that the time malus should be mitigated gradually and continuously to 100% as your proficiency approaches 100%. Furthermore, I'd like to propose using a sine-based function that models diminishing returns towards the end. Squaring it gives less benefits at low proficiency level.
![sin](https://user-images.githubusercontent.com/8000047/120942263-9f246780-c6dc-11eb-8c47-31fc2197a841.png)

On the other hand, the failure malus makes sense to be modeled as we discussed previously, with a sharper increase at the end and not continuous. You need to internalize all the critical lessons / techniques in order to fully eliminate the failure penalty. As long as there's one thing you haven't understood you can still screw up. Going from 99% to 100% proficient means learning that last thing you needed to learn. Jumping from 50% mitigation to 100% mitigation still seems a bit drastic to me (since you learned many things along on the way and I don't think the last bit of knowledge should matter quite that much). I'd like to propose 75% mitigation at 99.99% proficiency, but this is ultimately your call.
![quad2](https://user-images.githubusercontent.com/8000047/120942606-e7448980-c6de-11eb-9db2-c00accb26a6f.png)

I-am-Erk commented:
Yes, I really like your idea of a sigmoidal improvement to the time penalty. That also adds my favourite thing: subtle and easily maintained complexity that most players will hardly notice but that we know is there.

anothersimulacrum commented:
If you'd like to work on it now, you could cherry-pick the commits you need from https://github.com/CleverRaven/Cataclysm-DDA/pull/48650 , and then rebase onto 0.F-dev, then drop those commits when they get actually added to 0.F-dev.

eltank commented:
> If you'd like to work on it now, you could cherry-pick the commits you need from #48650 , and then rebase onto 0.F-dev, then drop those commits when they get actually added to 0.F-dev.

I'm probably going to regret this later but ok, that's done. Documenting the process in case it helps someone else, or future me:
- make a backup of the PR branch using git branch
- fetch 0-F.dev from upstream if necessary
- create a new temp branch from 0.F-dev
- cherrypick commits from dependency PRs into temp branch
- cherrypick commits from original PR into temp branch
- delete original PR branch
- re-create original PR branch as a clone of the temp branch
- delete temp branch
- edit the PR in Github to target 0.F-dev
- force push the PR branch to origin

This had the unfortunate side-effect of requesting reviews from lots of people. Sorry about that. Maybe I should have done the force-push before editing the PR.

anothersimulacrum commented:
> > If you'd like to work on it now, you could cherry-pick the commits you need from #48650 , and then rebase onto 0.F-dev, then drop those commits when they get actually added to 0.F-dev.
> 
> I'm probably going to regret this ~later~ [immediately] but ok, that's done. Documenting the process in case it helps someone else, or future me:
> 
>     * [optional] make a backup of the PR branch using git branch
> 
>     * fetch 0-F.dev from upstream if necessary
> 
>     * create a new temp branch from 0.F-dev
> 
>     * cherrypick commits from dependency PRs into temp branch
> 
>     * cherrypick commits from original PR into temp branch
> 
>     * delete original PR branch
> 
>     * re-create original PR branch as a clone of the temp branch
> 
>     * delete temp branch
> 
>     * edit the PR in Github to target 0.F-dev
> 
>     * force push the PR branch to origin
> 
> 
> This had the unfortunate side-effect of requesting reviews from lots of people. Sorry about that. Maybe I should have done the force-push before editing the PR.

I think it probably would have done it either way :(. I've removed the review requests now.

eltank commented:
> I think it probably would have done it either way :(. I've removed the review requests now.

Thanks!

>> If a helper NPC is more proficient than you, should their proficiency level be used in calculations instead of yours? If they're fully proficient, this is already the case (your missing proficiency is fully mitigated).
> I don't see why not. We can basically assume that they're doing it and showing you how if you're nonproficient, or giving you prompts and helping prepare things for you if you're partially proficient.

This is done.

I've also implemented the formulas discussed above in code.

Next I will get rid of the code duplication from `missing_proficiencies_string` with a bit of refactoring. I probably need to fix a broken test too.

anothersimulacrum commented on src/recipe.cpp:
```diff
@@ -729,32 +738,40 @@ std::set<proficiency_id> recipe::assist_proficiencies() const
 }
 
-float recipe::proficiency_time_maluses( const Character &guy ) const
+float recipe::proficiency_time_maluses( const Character &c ) const
```
Can we keep the non-single variable name?

eltank replied:
Sure, but can you suggest something else that's short and not gender-specific?
"Character *c" is used in 6 other functions in this file, so I was going for consistency. I could change it to "crafter" instead (as is the norm in crafting.cpp).

anothersimulacrum replied:
I don't really think it's worth the churn to change it. That said, that's because it hurts history, and you're already killing all the history with your functional changes. So, let's go with crafter.

anothersimulacrum commented on src/recipe.cpp:
```diff
@@ -729,32 +738,40 @@ std::set<proficiency_id> recipe::assist_proficiencies() const
 }
 
-float recipe::proficiency_failure_maluses( const Character &guy ) const
+float recipe::proficiency_failure_maluses( const Character &c ) const
```
Ditto.

anothersimulacrum commented on src/recipe.cpp:
```diff
@@ -729,32 +738,40 @@ std::set<proficiency_id> recipe::assist_proficiencies() const
-            float malus = 1.0f + ( prof.time_multiplier - 1.0f ) *
-                          ( 1.0f - guy.crafting_inventory().get_book_proficiency_bonuses().time_factor( prof.id ) );
-            total_malus *= malus;
+        if( !c.has_proficiency( prof.id ) &&
+            !helpers_have_proficiencies( c, prof.id ) && prof.time_multiplier > 1.0f ) {
+            double malus = prof.time_multiplier - 1.0;
+            malus *= 1.0 - c.crafting_inventory().get_book_proficiency_bonuses().time_factor( prof.id );
+            double pl = get_aided_proficiency_level( c, prof.id );
+            // Sigmoid function that mitigates 100% of the time malus as pl approaches 1.0
+            // but has little effect at pl < 0.5. See #49198
+            malus *= 1.0 - std::pow( 0.5 + 0.5 * std::sin( pl * M_PI - M_PI / 2 ), 2 );
+            total_malus *= 1.0 + malus;
```
Would it make sense to keep this formula contained to proficiency_set or a character function? e.g. would we want to use this formula somewhere else?

eltank replied:
This is reused, but only in this file. Please have another look post-refactor.

anothersimulacrum replied:
I'm asking more if we would want to use this function anywhere else. If we're just going to use it for crafting, it can live here, but if we're going to possibly use it more places, I think we could plan ahead.

eltank replied:
Well, I can't think of any reason why we'd want the penalty for a single proficiency other than for display purposes. What matters for simulation purposes is the total penalty taking into account all proficiencies, which is already exported as a public function.

anothersimulacrum commented on src/recipe.cpp:
```diff
@@ -729,32 +738,40 @@ std::set<proficiency_id> recipe::assist_proficiencies() const
-            float malus = 1.0f + ( prof.fail_multiplier - 1.0f ) *
-                          ( 1.0f - guy.crafting_inventory().get_book_proficiency_bonuses().fail_factor( prof.id ) );
-            total_malus *= malus;
+        if( !c.has_proficiency( prof.id ) &&
+            !helpers_have_proficiencies( c, prof.id ) && prof.fail_multiplier > 1.0f ) {
+            double malus =  prof.fail_multiplier - 1.0f;
+            malus *= 1.0 - c.crafting_inventory().get_book_proficiency_bonuses().fail_factor( prof.id );
+            double pl = get_aided_proficiency_level( c, prof.id );
+            // The failure malus is not completely eliminated until the proficiency is mastered.
+            // Most of the mitigation happens at higher pl. See #49198
+            malus *= 1.0 - ( 0.75 * std::pow( pl, 3 ) );
+            total_malus *= 1.0 + malus;
```
Would it make sense to keep this formula contained to proficiency_set or a character function? e.g. would we want to use this formula somewhere else?

anothersimulacrum commented on tests/crafting_test.cpp:
```diff
@@ -957,3 +924,35 @@ TEST_CASE( "book_proficiency_mitigation", "[crafting][proficiency]" )
+    GIVEN( "a recipe with required proficiencies" ) {
+        clear_avatar();
+        clear_map();
+        const recipe &test_recipe = recipe_id( "leather_belt" ).obj();
```
Exactly the same function, just a stylistic choice.
```suggestion
        const recipe &test_recipe = *recipe_id( "leather_belt" );
```

eltank replied:
Done.

anothersimulacrum commented on tests/crafting_test.cpp:
```diff
@@ -753,141 +753,111 @@ TEST_CASE( "crafting_skill_gain", "[skill],[crafting],[slow]" )
 {
     SECTION( "lvl 0 -> 1" ) {
         GIVEN( "nominal morale" ) {
-            test_skill_progression( recipe_id( "blanket" ), 175, 0, false );
```
Why are all these tests being removed?

eltank replied:
This is deliberate. Here is the reason:
The purpose of the test called "crafting_skill_gain" is to check how long it takes to gain a skill level via crafting and detect unintentional regression. When proficiencies were added this test started doing double-duty by also checking the impact of proficiencies. 
These tests are too brittle and will break any time there's a change in the proficiency penalty formulas or proficiency xp gain. This causes excessive maintenance burden since you have to update up to 36 expectations. A single such test should suffice for checking the impact of proficiency on crafting time. We should probably set up a new unit unit test dedicated to measuring how long it takes to level up proficiencies by simulating multiple crafting operations. But the more scenarios you simulate the more values need to be updated when you make any change to the formula -- a lot of work for little benefit.

jbytheway commented on src/recipe.cpp:
```diff
@@ -738,34 +696,95 @@ std::set<proficiency_id> recipe::assist_proficiencies() const
+        double pl = get_aided_proficiency_level( crafter, prof.id );
+        // Sigmoid function that mitigates 100% of the time malus as pl approaches 1.0
+        // but has little effect at pl < 0.5. See #49198
+        malus *= 1.0 - std::pow( 0.5 + 0.5 * std::sin( pl * M_PI - M_PI / 2 ), 2 );
```
You can simplify this

sin( πx - π/2 )
= -sin( π/2 - πx )
= -cos( πx )
```suggestion
        malus *= 1.0 - std::pow( 0.5 - 0.5 * std::cos( pl * M_PI ), 2 );
```
(although FWIW personally I think a convex curve makes more sense; I liked x^4/2)

eltank replied:
Thanks, done

actual-nh commented:
> I think it probably would have done it either way :(. I've removed the review requests now.

Not sure. It definitely does it if you do rebase+push first then change branch.

actual-nh commented:
So what is the current proposed curve for time and the current proposed curve for failure?

anothersimulacrum commented:
> > I think it probably would have done it either way :(. I've removed the review requests now.
> 
> Not sure. It definitely does it if you do rebase+push first then change branch.

Either way you pull in all the commits in the other branch that one branch lacks, I think it's just up to how close the master branch you're based on is to 0.F-dev.

eltank commented:
> So what is the current proposed curve for time and the current proposed curve for failure?

I've been keeping the PR description in sync with the code. Please check there.

actual-nh commented:
> > So what is the current proposed curve for time and the current proposed curve for failure?
> 
> I've been keeping the PR description in sync with the code. Please check there.

Whoops! Didn't see the changes in the PR until I did a forced reload.

eltank commented:
Now that 0.F-dev has been brought up to date, I have rebased this PR onto the latest 0.F-dev HEAD, so it's no longer in WIP state. I also took the opportunity to tidy up by squashing some commits.
I think Fris0uman got roped in by accident in this process.

eltank commented:
0.F is finally out and this PR has been rebased on master. Is there anything else I need to do?

anothersimulacrum commented on src/character.h:
```diff
@@ -1833,6 +1833,7 @@ class Character : public Creature, public visitable
 
         // --------------- Proficiency Stuff ----------------
         bool has_proficiency( const proficiency_id &prof ) const;
+        float get_proficiency_level( const proficiency_id &prof ) const;
```
Could you change this name from `_level` to `_practice` or similar? `level` has connotations that I don't think are quite correct here.

eltank replied:
Sure! Done.

actual-nh commented:
From #49779, there are some errors associated with this one?

anothersimulacrum commented:
The test case passes without running other ones, but because of focus (I think), the practice doesn't work (they only get 1 or 2% proficiency experience). I've got a fix for that compiling.

anothersimulacrum commented:
Fixed in https://github.com/CleverRaven/Cataclysm-DDA/pull/49779/commits/94f052a6bd490c9cb7b231962477796da88e4411
